pkg/kvstore/allocator/idpool.go

```diff
@@ -0,0 +1,281 @@
+** This is guaranteed never to occur.
+*/
+type idPool struct {
+	// mutex protects all idPool data structures
+	mutex lock.Mutex
+
+	// min is the lower limit when selecting IDs. The pool will never
+	// return an ID lesser than this value.
+	minID ID
+
+	// max is the upper limit when selecting IDs. The pool will never
+	// allocate an ID greater than this value.
+	maxID ID
+
+	// idCache is a cache of IDs backing the pool.
+	idCache *idCache
+
+	// Upon a refresh of the pool, idCache will be pointed to
+	// nextIDCache.
+	nextIDCache *idCache
+}
+
+func newIDPool(minID ID, maxID ID) *idPool {
+	p := &idPool{
+		minID: minID,
+		maxID: maxID,
+	}
+	p.StartRefresh()
+	p.FinishRefresh()
+
+	return p
+}
+
+// StartRefresh creates a new cache backing the pool.
+// This cache becomes live when FinishRefresh() is called.
+func (p *idPool) StartRefresh() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.nextIDCache = newIDCache(p.minID, p.maxID)
+}
+
+// FinishRefresh makes the most recent cache created by the pool live.
+func (p *idPool) FinishRefresh() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.idCache = p.nextIDCache
+}
+
+// SelectAvailableID returns an available ID at random from the pool.
+// that is leased to the caller.
+// Returns an ID of NoID if no there is no available ID in the pool.
+func (p *idPool) SelectAvailableID() ID {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.idCache.selectAvailableID()
+}
+
+// Release returns a leased ID back to the pool.
+// This operation accounts for IDs that were previously selected
+// from the pool but were unused, e.g if allocation was unsuccessful.
+// Thus, it has no effect if the ID is not currently leased in the
+// pool, or the pool has since been refreshed.
+//
+// Returns true if the ID was returned back to the pool as
+// a result of this call.
+func (p *idPool) Release(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.idCache.release(id)
+}
+
+// Use makes a leased ID unavailable in the pool and has no effect
+// otherwise. Returns true if the ID was made unavailable
+// as a result of this call.
+func (p *idPool) Use(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.idCache.use(id)
+}
+
+// Insert makes an unavailable ID available in the pool
+// and has no effect otherwise. Returns true if the ID
+// was added back to the pool.
+func (p *idPool) Insert(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.nextIDCache.insert(id)
+}
+
+// Remove makes an ID unavailable in the pool.
+// Returns true if the ID was previously available in the pool.
+func (p *idPool) Remove(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.nextIDCache.remove(id)
+}
+
+type idCache struct {
+
+	// ids is a slice of IDs available in this idCache.
+	ids []ID
+
+	// index tracks the position of IDs in the above ids slice.
+	index map[ID]int
+
+	// leased is a slice of IDs that are leased in this idCache.
+	leased map[ID]bool
```

Comment:
You don't take advantage of the fact that your map values are always `true`. So you might as well have no value, to make the intent clearer:
```
leased map[ID]struct{}
```